ako-gateway-api/k8s/validator.go

```diff
@@ -92,47 +104,81 @@ func IsValidGateway(key string, gateway *gatewayv1.Gateway) bool {
+		programmedCondition.
+			Reason(string(gatewayv1.GatewayReasonAddressNotUsable)).
+			SetIn(&gatewayStatus.Conditions)
+		_, err := akogatewayapistatus.Record(key, gateway, &akogatewayapistatus.Status{GatewayStatus: gatewayStatus})
+		if err != nil {
+			utils.AviLog.Errorf("key: %s, msg: Gateway status patch was not successful :%s ", key, err.Error())
+		}
 		return false
 	}
 
 	if len(spec.Addresses) == 1 && *spec.Addresses[0].Type != "IPAddress" {
 		utils.AviLog.Errorf("key: %s, msg: gateway address is not of type IPAddress %+v", key, gateway.Name)
 		defaultCondition.
+			Reason(string(gatewayv1.GatewayReasonUnsupportedAddress)).
 			Message("Only IPAddress as AddressType is supported").
 			SetIn(&gatewayStatus.Conditions)
-		akogatewayapistatus.Record(key, gateway, &akogatewayapistatus.Status{GatewayStatus: gatewayStatus})
+		programmedCondition.
+			Reason(string(gatewayv1.GatewayReasonAddressNotUsable)).
+			SetIn(&gatewayStatus.Conditions)
+		_, err := akogatewayapistatus.Record(key, gateway, &akogatewayapistatus.Status{GatewayStatus: gatewayStatus})
+		if err != nil {
+			utils.AviLog.Errorf("key: %s, msg: Gateway status patch was not successful :%s ", key, err.Error())
+		}
 		return false
 	}
 
 	gatewayStatus.Listeners = make([]gatewayv1.ListenerStatus, len(gateway.Spec.Listeners))
 
-	var invalidListenerCount int
+	var validListenerCount int
 	for index := range spec.Listeners {
-		if !isValidListener(key, gateway, gatewayStatus, index) {
-			invalidListenerCount++
+		if isValidListener(key, gateway, gatewayStatus, index) {
+			validListenerCount++
 		}
 	}
 
-	if invalidListenerCount > 0 {
-		utils.AviLog.Errorf("key: %s, msg: Gateway %s contains %d invalid listeners", key, gateway.Name, invalidListenerCount)
+	if validListenerCount == 0 {
+		utils.AviLog.Errorf("key: %s, msg: Gateway %s does not contains any valid listener", key, gateway.Name)
 		defaultCondition.
 			Type(string(gatewayv1.GatewayConditionAccepted)).
 			Reason(string(gatewayv1.GatewayReasonListenersNotValid)).
-			Message(fmt.Sprintf("Gateway contains %d invalid listener(s)", invalidListenerCount)).
+			Message("Gateway does not contain any valid listener").
+			SetIn(&gatewayStatus.Conditions)
+		programmedCondition.
 			SetIn(&gatewayStatus.Conditions)
 		akogatewayapistatus.Record(key, gateway, &akogatewayapistatus.Status{GatewayStatus: gatewayStatus})
+		gateway.Status = *gatewayStatus.DeepCopy()
 		return false
+	} else if validListenerCount < len(spec.Listeners) {
+		defaultCondition.
+			Reason(string(gatewayv1.GatewayReasonListenersNotValid)).
+			Message("Gateway contains atleast one valid listener").
+			SetIn(&gatewayStatus.Conditions)
+		_, err := akogatewayapistatus.Record(key, gateway, &akogatewayapistatus.Status{GatewayStatus: gatewayStatus})
+		if err == nil {
+			utils.AviLog.Infof("key: %s, msg: Gateway %s contains atleast one valid listener", key, gateway.Name)
```

Comment:
we can move this info log outside this condition block

Reply:
Done